src/main/java/com/google/devtools/build/lib/skyframe/ArtifactFunction.java

```diff
@@ -288,18 +290,18 @@ private static AggregatingArtifactValue createAggregatingValue(
         return null;
       }
       if (inputValue instanceof FileArtifactValue) {
-        inputs.add(Pair.of(input, (FileArtifactValue) inputValue));
+        fileInputs.add(Pair.of(input, (FileArtifactValue) inputValue));
       } else if (inputValue instanceof TreeArtifactValue) {
-        inputs.add(Pair.of(input, ((TreeArtifactValue) inputValue).getSelfData()));
+        directoryInputs.add(Pair.of(input, (TreeArtifactValue) inputValue));
```

Comment:
Does it work to simply put `((TreeArtifactValue) inputValue).getChildValues()` into the `inputs` list? Then, `AggregatingArtifactValue` wouldn't need to distinguish between directory and normal inputs.

Reply:
Yes, I guess it's unavoidable.